### DISCUSSION TITLE
[general] Should {nanonext} be made a 'Suggests' dependency?

### DISCUSSION THREAD

**Discussion by MilesMcBain** [ropensci/targets#1069] [General]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I noticed after failure to install dev `{targets}` that `{nanonext}` depends on `libnng` which is only [available for Ubuntu 20.04 and later](https://packages.ubuntu.com/search?keywords=libnng&searchon=names).

An environment I use is Ubuntu 18.04, which is not technically end of life until the end of this month. I do note that [Ubuntu 18.04 is supported out to 2028](https://ubuntu.com/blog/18-04-end-of-standard-support) with a pro subscription.

I will be doing my best to get our environment upgraded promptly, but I still think if it's a trivial change it might be worth making `{nanonext}` a suggested dependency to avoid catching people off guard.

**Comment by wlandau** [comment]
Moved to `Suggests:` in 451c30af85ca4fc4fe37a9883060e570308c4586.

I did a lot of work last week to reduce the overhead processing lots of targets, and part of that included replacing the slow `proc.time()` with the much faster `nanonext::mclock()` where possible. But as a `Suggests:` package, it turned out to be more complicated than I thought because a check of `requireNamespace("nanonext")` locally caused forked `clustermq`/`future` processes to crash. Today I did some more digging and found that `find.package()` is a nice alternative which does not cause these issues.

`nanonext` is required by `mirai`, which is required by [`crew`](https://wlandau.github.io/crew/). After the next round of package releases in a week or 2, [`crew`](https://wlandau.github.io/crew/) will supersede `clustermq` and `future` as the recommended way to run pipelines with parallel computing: https://books.ropensci.org/targets/crew.html

**Reply by MilesMcBain** [reply]
Great! Thanks for your hard work and responsiveness, as always Will.

I don't use any of this HPC stuff at the moment, but it's exciting to see {targets} gaining these superpowers.

**Reply by shikokuchuo** [reply]
@MilesMcBain I've just been made aware of this by @wlandau. Can you help me understand the issue you have installing {nanonext} on Ubuntu 18.04? The system library is only used if present and meets the minimum version requirement (> 1.5). Otherwise the bundled source will be automatically built, and all that requires is `cmake`. It is not a heavy dependency at all - the entire package is only ~800k. There is a much more substantive use that would benefit {targets} I have been working with @wlandau on in this thread https://github.com/ropensci/targets/issues/1139.

**Comment by MilesMcBain** [comment]
@shikokuchuo 

Well I don’t have the problem any longer since we upgraded Ubuntu as I noted we would in the initial issue.

I can’t remember the error clearly but dev targets could not be successfully installed, and failed with some kind of compilation error, which ended up being due to the system missing libnng. Again as I noted in the original issue.


Keep in mind the timing of this was shortly after nanonext was added as a dependency to {targets} so I am not sure if the conditional vendoring of code you describe was in play then?

**Reply by shikokuchuo** [reply]
If I understand correctly then, this is not a case of you being forced to use a system library from your package manager (which does not exist), but just that the compilation failed at the time.

If so then I am fairly certain that this was caused by this bug shikokuchuo/nanonext#16, which was introduced by 0.8.3, released just a few days before you reported your initial problem, and now fixed in 0.10.1.

I don't see any inherent problem in installing on old Ubuntu versions. Thanks for your help in understanding this.

**Comment by MilesMcBain** [comment]
It's possible it was caused by https://github.com/shikokuchuo/nanonext/issues/16, but the error text in that report does not look familiar.

As I recall the error was much more like the standard stuff you get when you're missing things on linux, I have a feeling it was a missing header file.

**Reply by shikokuchuo** [reply]
Sure, we don't know what the exact code was that caused the initial failure. But bigger picture, there is no problem _per se_ with 'libnng' not being available in your package manager, given that it would build from the bundled source as I have described.
